firestore.rules

```diff
@@ -80,15 +107,45 @@ service cloud.firestore {
     match /authed/{portal} {
       allow read, write: if isAuthedTeacher();
 
+      match /documents/{docId} {
+        // portal-authenticated teachers can create valid documents
+        allow create: if isAuthedTeacher() && isValidDocumentRequest();
+        // teachers can only update their own documents and only if they're valid
+        allow update: if isAuthedTeacher() && isValidDocumentRequest() && resourceMatchUserId();
+        // teachers can only delete their own documents
+        allow delete: if isAuthedTeacher() && resourceMatchUserId();
+        // only teachers that "own" the document can read it (for now)
+        allow read: if isAuthedTeacher() && resourceMatchTeacher();
+
+        function getDocumentTeachers() {
+          return get(/databases/$(database)/documents/authed/$(portal)/documents/$(docId)).data.teachers;
+        }
+
+        function teacherInDocumentTeachers() {
+          return isAuthedTeacher() && (request.auth.uid in getDocumentTeachers());
+        }
+
+        match /comments/{commentId} {
+          // portal-authenticated teachers with access to the document can create valid comments
+          allow create: if teacherInDocumentTeachers() && isValidCommentRequest();
+          // teachers can only update their own comments and only if they're valid
+          allow update: if teacherInDocumentTeachers() && isValidCommentRequest() && resourceMatchUserId();
```

Comment:
Some issue here I think. A user could change the uid of the comment to be some someone other than themselves.

Reply:
Good point -- I've updated the rules to require that updates don't change read-only fields.